### PR TITLE
feat: v2 DUP alerts backend

### PR DESCRIPTION
**Pull request by jzimbel-mbta**
**Asana task**: [[DUP v2] Alerts Backend](https://app.asana.com/0/1185117109217413/1203830054341500/f)

I needed to deviate from the [tech spec](https://github.com/mbta/screens/blob/master/docs/tech_specs/dup_v2_alert_widget_backend.md) somewhat in order to make the new widget struct compatible with the `BaseAlert` functions. Otherwise I would have needed to duplicate much of that logic.

This also meant I needed to use the same alert config as other widgets, with just a `stop_id` field. That's fine though, because the more complicated config that old DUP alerts use was actually unnecessary and just a consequence of them reusing parts of the DUP departures config.

Remaining work:
- [x] Finish filling in the more complicated/edge-case-laden serialization logic in the `DupAlert.Serialize` module
- [ ] Finish unit tests
- [x] Merge with latest master
- [x] Have `BaseAlert` access alert widget info via protocol callbacks only, instead of directly accessing struct fields by name with no up-front way to tell what those required fields are

---
- [ ] Tests added?

### CODE REVIEW

File: lib/screens/v2/widget_instance/common/base_alert.ex [hunk 1]
```diff
@@ -231,17 +231,9 @@ defmodule Screens.V2.WidgetInstance.Common.BaseAlert do
...
-
-            RouteType.from_id(route_type_id) == rt ->
-              {:halt, empty_set}
-
-            true ->
-              {:cont, uninformed}
-          end
+          if RouteType.from_id(route_type_id) in rts,
+            do: {:halt, empty_set},
+            else: {:cont, uninformed}
```
Comment (hannahpurcell) [inline lib/screens/v2/widget_instance/common/base_alert.ex:236]: Nice clean-up in this file

**Review by hannahpurcell — COMMENTED** [review]
Partial review, since I still have 2 more files to check out

File: lib/screens/v2/widget_instance/common/base_alert.ex [hunk 2]
```diff
@@ -9,12 +11,13 @@ defmodule Screens.V2.WidgetInstance.Common.BaseAlert do
...
   alias Screens.V2.WidgetInstance.Alert, as: AlertWidget
   alias Screens.V2.WidgetInstance.ReconstructedAlert
+  alias Screens.V2.WidgetInstance.DupAlert
 
   @type stop_id :: String.t()
 
   @type route_id :: String.t()
 
-  @spec home_stop_id(AlertWidget.t() | ReconstructedAlert.t()) :: String.t()
+  @spec home_stop_id(AlertWidget.t() | ReconstructedAlert.t() | DupAlert.t()) :: String.t()
```
Comment (hannahpurcell) [inline lib/screens/v2/widget_instance/common/base_alert.ex:20]: Line 24 needs to include `DUP`

File: lib/screens/v2/widget_instance/dup_alert.ex [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    }
+  end
+
+  @spec serialize_partial(t()) :: %{
+          alert_text: %FreeTextLine{icon: :warning, text: list(FreeText.t())},
+          color: :red | :orange | :green | :blue
+        }
+  defp serialize_partial(t) do
+    text = Serialize.partial_alert_free_text(t)
+    color = Serialize.route_color(t)
```
Comment (hannahpurcell) [inline lib/screens/v2/widget_instance/dup_alert.ex:120]: Functions / files do not exist error

**Review by hannahpurcell — CHANGES_REQUESTED** [review]
Pausing review for now while it's still WIP. (Not yet able to view the serialized alert)

File: lib/screens/v2/candidate_generator/dup.ex [hunk 4]
```diff
@@ -92,6 +103,125 @@ defmodule Screens.V2.CandidateGenerator.Dup do
...
 
+  def alert_instances(config, now \\ DateTime.utc_now()) do
+    # In this function:
+    # - Fetch relevant alerts for all SUBWAY/LIGHT RAIL routes serving this stop
+    # - Check for special cases. If there is one, just use that. Otherwise:
+    # - Select one alert
+    # - Create 3 candidate structs from the alert, one for each rotation
+    %Screen{app_params: %Dup{alerts: %AlertsConfig{stop_id: stop_id}}} = config
+
+    stop_name = Stop.fetch_stop_name(config.app_params.header.stop_id)
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/candidate_generator/dup.ex:114]: Just for consistency—I don't think we would, but if we configure the header widget with a different stop ID, we wouldn't want the rendered stop name to be different (due to reading stop ID from the alerts config) when the header happens to be rendered as part of the alert widget during a full-screen alert.

File: lib/screens/v2/widget_instance/common/base_alert.ex (hunk 2 above)
Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/common/base_alert.ex:20]: Pushing my changes related to this shortly--all of the values that BaseAlert depends on have been laid out in a new `SingleAlertWidget` protocol now, so these functions all just operate on `SingleAlertWidget.t()` instead of the union of `AlertWidget.t()`, `ReconstructedAlert.t()`, `DupAlert.t()`.

File: lib/screens/dup_screen_data.ex [hunk 5]
```diff
@@ -144,7 +144,7 @@ defmodule Screens.DupScreenData do
 
   defp fetch_and_interpret_alerts(%Dup.Departures{sections: sections}) do
     sections
-    |> Task.async_stream(&fetch_and_interpret_alert/1, timeout: :infinity)
+    |> Task.async_stream(&fetch_and_interpret_alert/1, timeout: 30_000)
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/dup_screen_data.ex:147]: Me being scatterbrained again, sorry. Paul S recommended this like a year ago, and it seemed too small for a full eng task, so when I was working on / reading code around these `Task.async_stream` calls, I just went ahead and made the change.

Happy to pull this out to a separate PR to reduce clutter.

**Review by hannahpurcell — COMMENTED** [review]
This should be broken down into separate PRs so the diff is more like 10 or less files significantly changed. (Slight variation is ok.)

Also, I think the hookup between backend and frontend wasn't quite right. Made an RL boundary alert for Broadway, and it's having React errors. I'll explore fixing in my branch so I can work on the special cases - will let ya know if it's straightforward.

File: lib/screens/v2/candidate_generator/dup.ex [hunk 6]
```diff
@@ -92,6 +103,125 @@ defmodule Screens.V2.CandidateGenerator.Dup do
...
+    # Special cases go here!
+    # Use the DupSpecialCaseAlert widget to directly specify the serialized data.
+
+    {:normal, alerts}
+  end
+
+  @spec choose_alert(list(Alert.t())) :: Alert.t() | nil
+  defp choose_alert([alert]), do: alert
+
+  defp choose_alert(alerts) do
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/candidate_generator/dup.ex]: ✅

File: lib/screens/v2/candidate_generator/dup.ex [hunk 7]
```diff
@@ -92,6 +103,125 @@ defmodule Screens.V2.CandidateGenerator.Dup do
...
+    # In this function:
+    # - Fetch relevant alerts for all SUBWAY/LIGHT RAIL routes serving this stop
+    # - Check for special cases. If there is one, just use that. Otherwise:
+    # - Select one alert
+    # - Create 3 candidate structs from the alert, one for each rotation
+    %Screen{app_params: %Dup{alerts: %AlertsConfig{stop_id: stop_id}}} = config
+
+    stop_name = Stop.fetch_stop_name(config.app_params.header.stop_id)
+
+    route_type_filter = Enum.map([:light_rail, :subway], &Screens.RouteType.to_id/1)
```
[the remark this answers is not in the document]
Comment (hannahpurcell) [reply lib/screens/v2/candidate_generator/dup.ex]: Oh derp, I clearly wasn't thinking about the special cases work when I mentioned WTC. Thanks for asking about it in slack. 👍

File: lib/screens/v2/widget_instance/dup_alert.ex [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+  defp routes_to_lines(route_ids) do
+    route_ids
+    |> Enum.map(fn
+      "Green-" <> _branch -> "Green"
+      other_subway_route -> other_subway_route
+    end)
+    |> Enum.uniq()
+  end
+
+  def alert_ids(%__MODULE__{} = t), do: t.alert.id
```
Comment (hannahpurcell) [inline lib/screens/v2/widget_instance/dup_alert.ex:244]: Gives an `Screens.ScreensByAlert.SelfRefreshRunner` error. Should be:

```suggestion
  def alert_ids(%__MODULE__{} = t), do: [t.alert.id]
```

File: lib/screens/config/v2/free_text/free_text_line.ex [hunk 9]
```diff
@@ -26,6 +26,8 @@ defmodule Screens.Config.V2.FreeTextLine do
           | :green_e
           | nil
 
+  @derive Jason.Encoder
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/config/v2/free_text/free_text_line.ex:29]: It makes it so `Jason` knows how to turn the struct into a chunk of JSON for the response! Basically it makes `Jason` treat it like a regular map. That way, we can put `FreeTextLine` structs in the maps returned by widgets' `serialize` functions without first converting them to maps.

[docs 1](https://hexdocs.pm/jason/readme.html#differences-to-poison)
> no support for encoding arbitrary structs - explicit implementation of the [Jason.Encoder](https://hexdocs.pm/jason/Jason.Encoder.html) protocol is always required.

[docs 2](https://hexdocs.pm/jason/Jason.Encoder.html)

[Also used in some of the elevator status widget's serialized data.](https://github.com/mbta/screens/blob/master/lib/screens/v2/widget_instance/elevator_status.ex#L17)

File: lib/screens/v2/candidate_generator/dup.ex [hunk 10]
```diff
@@ -107,8 +112,6 @@ defmodule Screens.V2.CandidateGenerator.Dup do
     List.duplicate(%NormalHeader{screen: config, icon: :logo, text: stop_name, time: now}, 3)
   end
 
-  ### End Header
-
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/candidate_generator/dup.ex:111]: Oops thanks, totally missed that

File: lib/screens/v2/widget_instance/dup_alert.ex [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+
+    Logger.info(
+      "[DUP alert no matching layout] " <>
+        Enum.map_join(log_fields, " ", fn {label, value} -> "#{label}=#{value}" end)
+    )
+  end
+
+  # Inputs/outputs are stored as a table for readability.
+  # Table is adapted from the one in the product spec: https://www.notion.so/mbta-downtown-crossing/DUP-Alert-Widget-Specification-a82acff850ed4f2eb98a04e5f3e0fe52
+  # Compile-time code converts each table row to `do_alert_layout` function clauses.
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/dup_alert.ex:140]: I agree it turned out a bit more brittle than I wanted when I started out, since it's not easy to generate the appropriate clauses for values of "any" or "any boundary" in the table, forcing me to use explicit location values and duplicate rows.

However I still prefer keeping the conditions in a table like this for conciseness and readability. While working on this, I was able to be confident that the code follows the spec because the logic is encoded here almost exactly the same as it is in the spec.

> Let's say we had new cases where a particular station is an exception, so we want a column with station name for only that one case. We could easily do that with a new pattern matching function, but then we'll have some cases in the table, and some cases written separately.

This is true, but the `delay` cases are already handled separately with a hand-written clause. Not necessarily a good thing, just saying this already happens. Maybe it would be enough to add a "comment" line to the table for each special case that's handled separately?

File: lib/screens/v2/widget_instance/dup_alert/serialize.ex [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    police_action: "police action",
+    power_problem: "a power issue",
+    signal_problem: "a signal problem",
+    snow: "snow conditions",
+    special_event: "a special event",
+    switch_problem: "a switch problem",
+    track_problem: "a track problem",
+    traffic: "traffic",
+    weather: "weather conditions"
+  }
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/dup_alert/serialize.ex:176]: Copied and adapted from DupScreenData.Response; I think ReconstructedAlert did the same! 😅

I agree that it should probably live in one place instead of 3, lol. I'll find a spot for it.

File: lib/screens/v2/widget_instance/dup_alert.ex [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+
+  require Logger
+
+  @enforce_keys [
+    :screen,
+    :alert,
+    :stop_sequences,
+    :subway_routes_at_stop,
+    :rotation_index,
+    :stop_name
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/dup_alert.ex:19]: Used [here!](https://github.com/mbta/screens/pull/1634/files#diff-bdbac4f3f941494c2eeead7667ee6ec6cbab0e7e0e3c0da670ca04cb2f742b50R30)

File: lib/screens/v2/widget_instance/dup_alert/serialize.ex [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+      %{format: :bold, text: unquote(str)}
+    end
+  end
+
+  # Provides a pattern-matchable shorthand for shrinking some text
+  defmacrop small(str) do
+    quote do
+      %{format: :small, text: unquote(str)}
+    end
+  end
```
[the remark this answers is not in the document]
Comment (hannahpurcell) [reply lib/screens/v2/widget_instance/dup_alert/serialize.ex:56]: Sure, just don't want to lose track of this. Mention these functions in the task, or leave a comment here indicating the future "to-do" for moving them, or go ahead and make a util file where you want the helpers to go eventually, but do not do the refactor.

File: lib/screens/alerts/alert.ex [hunk 15]
```diff
@@ -269,14 +272,22 @@ defmodule Screens.Alerts.Alert do
...
   end
 
-  defp format_query_param({:route_types, route_types}) do
-    route_type_ids = Enum.map_join(route_types, ",", &Screens.RouteType.to_id/1)
+  defp format_query_param({:route_types, route_types}) when is_list(route_types) do
+    route_type_ids =
+      Enum.map_join(route_types, ",", fn
+        rt when RouteType.is_route_type(rt) -> RouteType.to_id(rt)
+        id when RouteType.is_route_type_id(id) -> id
+      end)
```
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/alerts/alert.ex:280]: Yeah I think I went a little overboard here. I'll go back to just accepting atoms, now with the guard to enforce it.

File: lib/screens/v2/widget_instance/dup_alert.ex (hunk 11 above)
[the remark this answers is not in the document]
Comment (jzimbel-mbta) [reply lib/screens/v2/widget_instance/dup_alert.ex:140]: Sorry, I meant the `delay` case was already mentioned in a comment ([here](https://github.com/mbta/screens/pull/1634/files#diff-97879ae25268a9f3b56c913f936c95193ad09831de6b1b5b764c35e74372f7eaR162)), I was just talking about what we'd do for your hypothetical case where we'd need to handle a specific exceptional station.

File: lib/screens/config/v2/dup.ex [hunk 16]
```diff
@@ -2,25 +2,26 @@ defmodule Screens.Config.V2.Dup do
...
+  alias Screens.Config.V2.{Alerts, Departures, EvergreenContentItem}
 
   @type t :: %__MODULE__{
           header: CurrentStopId.t() | CurrentStopName.t(),
           evergreen_content: list(EvergreenContentItem.t()),
           primary_departures: Departures.t(),
-          secondary_departures: Departures.t()
+          secondary_departures: Departures.t(),
+          alerts: Alerts.t(),
+          evergreen_content: list(EvergreenContentItem.t())
```
Comment (hannahpurcell) [inline lib/screens/config/v2/dup.ex:13]: merge weirdness

```suggestion
          alerts: Alerts.t()
```

**Review by hannahpurcell — APPROVED** [review]
Yay!!! One fix noted above. Otherwise, looks great. Thanks, Jon

File: lib/screens/config/v2/dup.ex (hunk 16 above)
Comment (jzimbel-mbta) [reply lib/screens/config/v2/dup.ex:13]: Thanks for catching that! I'm sure it's a mistake I made while addressing merge conflicts 😵‍💫
